File: leetcode/2685/c++/solution.cpp
/*
    2685. Count the Number of Complete Components
    Author: Miguel Angel Bermeo Ayerbe
*/
class Solution {
   public:
    int countCompleteComponents(int n, vector<vector<int>>& edges) {
        std::vector<std::vector<size_t>> graph(n, std::vector<size_t>());
        std::vector<std::bitset<50>> matrix_graph(n, std::bitset<50>());
        for (const std::vector<int>& edge_i : edges) {
            graph[edge_i[0]].push_back(edge_i[1]);
            graph[edge_i[1]].push_back(edge_i[0]);
            matrix_graph[edge_i[0]][edge_i[1]] = true;
            matrix_graph[edge_i[1]][edge_i[0]] = true;
        }
        std::bitset<50> marks, temp;
        int output = 0, is_ccc = 0;
        for (size_t i = 0; i < n; ++i) {
            if (!marks[i]) {
                temp = matrix_graph[i];
                temp[i] = true;
                is_ccc = 1;
                for (const size_t& node_i : graph[i]) {
                    matrix_graph[node_i][node_i] = true;
                    if (temp != matrix_graph[node_i]) {
                        is_ccc = 0;
                    }
                    temp |= matrix_graph[node_i];
                }
                matrix_graph[i][i] = true;
                output += is_ccc;
                marks |= temp;
            }
        }
        return output;
    }
};

File: leetcode/2685/rust/solution.rs
/*
    2685. Count the Number of Complete Components
    Author: Miguel Angel Bermeo Ayerbe
*/
use std::collections::HashMap;
impl Solution {
    pub fn find_parent(parents: &mut Vec<usize>, id_node: usize) -> usize {
        let mut id_parent: usize = parents[id_node];
        if id_parent != parents[id_parent] {
            id_parent = Self::find_parent(parents, id_parent)
        }
        id_parent
    }

    pub fn union_nodes(
        parents: &mut Vec<usize>,
        ranks: &mut Vec<usize>,
        id_node_1: usize,
        id_node_2: usize,
    ) {
        let id_parent_1 = Self::find_parent(parents, id_node_1);
        let id_parent_2 = Self::find_parent(parents, id_node_2);
        if id_parent_1 != id_parent_2 {
            if ranks[id_parent_1] < ranks[id_parent_2] {
                parents[id_parent_1] = id_parent_2;
            } else if ranks[id_parent_1] > ranks[id_parent_2] {
                parents[id_parent_2] = id_parent_1;
            } else {
                parents[id_parent_2] = id_parent_1;
                ranks[id_parent_1] += 1;
            }
        }
    }

    pub fn count_complete_components(n: i32, edges: Vec<Vec<i32>>) -> i32 {
        let mut output: i32 = 0;
        let n_nodes = n as usize;
        let mut parents: Vec<usize> = (0..n_nodes).collect();
        let mut ranks: Vec<usize> = vec![0; n_nodes];
        // 1. Union edges
        for edge_i in &edges {
            Self::union_nodes(
                &mut parents,
                &mut ranks,
                edge_i[0] as usize,
                edge_i[1] as usize,
            );
        }
        // 2. Count number of nodes
        let mut map: HashMap<usize, (usize, usize)> = HashMap::new();
        for id_node in 0..n_nodes {
            map.entry(Self::find_parent(&mut parents, id_node))
                .or_insert((0, 0))
                .0 += 1;
        }
        // 3. Count number of edges
        for edge_i in &edges {
            map.entry(Self::find_parent(&mut parents, edge_i[0] as usize))
                .or_insert((0, 0))
                .1 += 1;
        }
        // 4. Compute complete components
        for (root, (num_nodes, num_edges)) in map.iter() {
            if *num_edges == (*num_nodes * (*num_nodes - 1)) / 2 {
                output += 1;
            }
        }
        output
    }
}
